pageserver/src/http/routes.rs

```diff
@@ -743,22 +746,41 @@ async fn get_lsn_by_timestamp_handler(
     let result = timeline
         .find_lsn_for_timestamp(timestamp_pg, &cancel, &ctx)
         .await?;
+
     #[derive(serde::Serialize, Debug)]
     struct Result {
         lsn: Lsn,
         kind: &'static str,
+        #[serde(default)]
+        #[serde(skip_serializing_if = "Option::is_none")]
+        #[serde(flatten)]
+        lease: Option<LsnLease>,
     }
     let (lsn, kind) = match result {
         LsnForTimestamp::Present(lsn) => (lsn, "present"),
         LsnForTimestamp::Future(lsn) => (lsn, "future"),
         LsnForTimestamp::Past(lsn) => (lsn, "past"),
         LsnForTimestamp::NoData(lsn) => (lsn, "nodata"),
     };
-    let result = Result { lsn, kind };
+
+    let lease = if with_lease && (kind == "present" || kind == "future") {
+        timeline
+            .make_lsn_lease(lsn, timeline.get_lsn_lease_length_for_ts(), &ctx)
+            .ok()
```

Comment:
Should we log errors here?

Reply:
>  I also peeked at the lease implementation and was surprised the leases are not persisted to disk anywhere. Are we ok with losing them on restart?

Based on discussion with @jcsp, we made leases in-memory only. To avoid any issues after restart, we [wait for lease period before starting GC after restart](https://github.com/neondatabase/neon/blob/yuchen/get-lsn-by-ts-leases/pageserver/src/tenant/tasks.rs#L367) instead of persisting it to disk. After waiting lease period, all leases that were granted before restart will expire.

Reply:
> Should we log errors here?

I think we should log error. However, I feel even if we cannot make a lease, we should not fail the get lsn request? It will just not have a `valid_until` field in the response.

Reply:
I think it makes sense to go through with the request as you're suggesting. The caller can retry if a lease is required.